// visual-toolkit/dist/themes/index.js
/**
 * THEME REGISTRATION SYSTEM
 *
 * Allows registering custom themes and accessing built-in themes.
 */
const registeredThemes = {};
/**
 * Register a custom theme
 *
 * @example
 * registerTheme('jungle', {
 *   colors: { canopy: { light: '#2d5016', mid: '#1a3009' } },
 *   particles: { rain: (count) => createRainParticles(count) },
 * });
 */
export function registerTheme(name, theme) {
    if (registeredThemes[name]) {
        console.warn(`Theme "${name}" is already registered. Overwriting.`);
    }
    registeredThemes[name] = theme;
}
/**
 * Get a registered theme
 */
export function getTheme(name) {
    return registeredThemes[name];
}
/**
 * Get all registered theme names
 */
export function getThemeNames() {
    return Object.keys(registeredThemes);
}
/**
 * Themes object - access registered themes
 *
 * @example
 * import { themes } from 'visual-toolkit';
 * themes.deepSea.colors.background
 * themes.jungle.particles.rain(50)
 */
export const themes = new Proxy({}, {
    get(target, prop) {
        if (typeof prop !== 'string')
            return undefined;
        return registeredThemes[prop];
    },
    ownKeys() {
        return Object.keys(registeredThemes);
    },
    has(target, prop) {
        return prop in registeredThemes;
    },
});
