// visual-toolkit/dist/scenes/utilities/animation-loop.js
/**
 * Animation Loop Manager
 *
 * Handles requestAnimationFrame lifecycle with frame counting
 * and delta time calculation
 */
export class AnimationLoop {
    constructor(onFrame) {
        this.onFrame = onFrame;
        this.animationId = null;
        this.frameCount = 0;
        this.lastFrameTime = 0;
        this.isRunning = false;
        this.render = () => {
            if (!this.isRunning)
                return;
            const now = performance.now();
            const deltaTime = now - this.lastFrameTime;
            this.lastFrameTime = now;
            this.onFrame(this.frameCount, deltaTime);
            this.frameCount++;
            this.animationId = requestAnimationFrame(this.render);
        };
    }
    /**
     * Start animation loop
     */
    start() {
        if (this.isRunning)
            return;
        this.isRunning = true;
        this.lastFrameTime = performance.now();
        this.frameCount = 0;
        this.render();
    }
    /**
     * Stop animation loop
     */
    stop() {
        if (this.animationId !== null) {
            cancelAnimationFrame(this.animationId);
            this.animationId = null;
        }
        this.isRunning = false;
    }
    /**
     * Reset frame counter
     */
    reset() {
        this.frameCount = 0;
        this.lastFrameTime = performance.now();
    }
    /**
     * Get current frame number
     */
    getFrameNum() {
        return this.frameCount;
    }
    /**
     * Check if animation is running
     */
    isRunning_() {
        return this.isRunning;
    }
}
