// visual-toolkit/dist/scenes/utilities/responsive-canvas.js
/**
 * Responsive Canvas Manager
 *
 * Handles canvas resizing to match container and maintains
 * aspect ratio while notifying observers of size changes
 */
export class ResponsiveCanvas {
    constructor(canvas) {
        this.resizeObserver = null;
        this.canvas = canvas;
    }
    /**
     * Start responsive resizing
     *
     * Canvas will automatically resize to match container
     */
    startResizing() {
        if (this.resizeObserver)
            return; // Already started
        const resize = this.handleResize.bind(this);
        this.resizeObserver = new ResizeObserver(resize);
        this.handleResize(); // Initial resize
        this.resizeObserver.observe(this.canvas.parentElement);
    }
    /**
     * Stop responsive resizing
     */
    stopResizing() {
        if (this.resizeObserver) {
            this.resizeObserver.disconnect();
            this.resizeObserver = null;
        }
    }
    /**
     * Register callback for resize events
     */
    registerResizeCallback(callback) {
        this.resizeCallback = callback;
    }
    /**
     * Get current canvas size
     */
    getSize() {
        return {
            width: this.canvas.width,
            height: this.canvas.height
        };
    }
    /**
     * Get canvas center
     */
    getCenter() {
        return {
            x: this.canvas.width / 2,
            y: this.canvas.height / 2
        };
    }
    /**
     * Get canvas bounds
     */
    getBounds() {
        return this.canvas.getBoundingClientRect();
    }
    handleResize() {
        const rect = this.canvas.getBoundingClientRect();
        const dpr = window.devicePixelRatio || 1;
        // Set canvas resolution
        this.canvas.width = rect.width * dpr;
        this.canvas.height = rect.height * dpr;
        // Scale canvas context to match DPR
        const ctx = this.canvas.getContext('2d');
        if (ctx) {
            ctx.scale(dpr, dpr);
        }
        // Notify observers
        if (this.resizeCallback) {
            this.resizeCallback();
        }
    }
}
